backend/src/middleware/filterBubble/replaceParams.spec.js

```diff
@@ -0,0 +1,129 @@
+import replaceParams from './replaceParams.js'
+
+describe('replaceParams', () => {
+  let args
+  let context
+  let run
+
+  let action = () => {
+    return replaceParams(args, context)
+  }
+
+  beforeEach(() => {
+    args = {}
+    run = jest.fn().mockResolvedValue({
+      records: [{ get: () => 1 }, { get: () => 2 }, { get: () => 3 }],
+    })
+    context = {
+      driver: {
+        session: () => {
+          return {
+            run,
+            close: () => {},
+          }
+        },
+      },
+    }
+  })
+
+  describe('args == ', () => {
+    describe('{}', () => {
+      it('does not crash', async () => {
+        await expect(action()).resolves.toEqual({})
+      })
+    })
+
+    describe('unauthenticated user', () => {
+      beforeEach(() => {
+        context.user = null
+      })
+
+      describe('{ filterBubble: { author: following } }', () => {
+        it('throws error', async () => {
+          args = { filterBubble: { author: 'following' } }
+          await expect(action()).rejects.toThrow('You are unauthenticated')
+        })
+      })
+
+      describe('{ filterBubble: { author: all } }', () => {
+        it('removes filterBubble param', async () => {
+          const expected = {}
+          await expect(action()).resolves.toEqual(expected)
+        })
+
+        it('does not make database calls', async () => {
+          await action()
+          expect(run).not.toHaveBeenCalled()
+        })
+      })
+    })
+
+    describe('authenticated user', () => {
+      beforeEach(() => {
+        context.user = { id: 'u4711' }
+      })
+
+      describe('{ filterBubble: { author: following } }', () => {
+        beforeEach(() => {
+          args = { filterBubble: { author: 'following' } }
+        })
+
+        it('returns args object with resolved ids of followed users', async () => {
+          const expected = { filter: { author: { id_in: [1, 2, 3] } } }
+          await expect(action()).resolves.toEqual(expected)
+        })
+
+        it('makes database calls', async () => {
+          await action()
+          expect(run).toHaveBeenCalledTimes(1)
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-04T11:32:21Z" title="Tuesday, June 4th 2019, 1:32:21 pm +02:00">Jun 4, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,129 @@
+import replaceParams from './replaceParams.js'
+
+describe('replaceParams', () => {
+  let args
+  let context
+  let run
+
+  let action = () => {
+    return replaceParams(args, context)
+  }
+
+  beforeEach(() => {
+    args = {}
+    run = jest.fn().mockResolvedValue({
+      records: [{ get: () => 1 }, { get: () => 2 }, { get: () => 3 }],
+    })
+    context = {
+      driver: {
+        session: () => {
+          return {
+            run,
+            close: () => {},
+          }
+        },
+      },
+    }
+  })
+
+  describe('args == ', () => {
+    describe('{}', () => {
+      it('does not crash', async () => {
+        await expect(action()).resolves.toEqual({})
+      })
+    })
+
+    describe('unauthenticated user', () => {
+      beforeEach(() => {
+        context.user = null
+      })
+
+      describe('{ filterBubble: { author: followed } }', () => {
+        it('throws error', async () => {
+          args = { filterBubble: { author: 'followed' } }
+          await expect(action()).rejects.toThrow('You are unauthenticated')
+        })
+      })
+
+      describe('{ filterBubble: { author: all } }', () => {
+        it('removes filterBubble param', async () => {
+          const expected = {}
+          await expect(action()).resolves.toEqual(expected)
+        })
+
+        it('does not make database calls', async () => {
+          await action()
+          expect(run).not.toHaveBeenCalled()
+        })
+      })
+    })
+
+    describe('authenticated user', () => {
+      beforeEach(() => {
+        context.user = { id: 'u4711' }
+      })
+
+      describe('{ filterBubble: { author: followed } }', () => {
+        beforeEach(() => {
+          args = { filterBubble: { author: 'followed' } }
+        })
+
+        it('returns args object with resolved ids of followed users', async () => {
+          const expected = { filter: { author: { id_in: [1, 2, 3] } } }
+          await expect(action()).resolves.toEqual(expected)
+        })
+
+        it('makes database calls', async () => {
+          await action()
+          expect(run).toHaveBeenCalled()
```


Is there any benefit here of using `toHaveBeenCalledTimes(1)` or `toHaveBeenCalledWith('whatever args')`
I typically try to be more specific because I want the test to fail if it is called more than once, or with the wrong args

One use cause for this that I saw with @aonomike, is that if you don't properly `clearAllMocks()` in between tests, you can get false positives.

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-04T13:06:36Z" title="Tuesday, June 4th 2019, 3:06:36 pm +02:00">Jun 4, 2019</time>_

---


@mattwr18 I usually don't try to be overly specific, so you don't need to refactor the tests if you refactor the code. What about `clearAllMocks`? I didn't know that. What is the exact problem? Does it happen even if you re-initialize the variables in a `beforeEach` block?

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-04T17:20:57Z" title="Tuesday, June 4th 2019, 7:20:57 pm +02:00">Jun 4, 2019</time>_

---


The problem is that unless you clear the mocks' calls one of your previous tests can call your mock function and then you get a false positive cause you think the it block that you are trying to test has called the function, but it's really a previous test.

I would favor using `toHaveBeenCalledTimes(1)` because there's not really a use case I can think of where you would want it to be called multiple times, but if you prefer to leave it as `toHaveBeenCalled()` or either way really, I think it's better to have an afterEach, which calls `jest.clearAllMocks()` to ensure you are testing what you think you are.

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-05T16:38:18Z" title="Wednesday, June 5th 2019, 6:38:18 pm +02:00">Jun 5, 2019</time>_

---


done!